aten/src/ATen/native/mkldnn/MKLDNNTensorImpl.cpp

```diff
@@ -0,0 +1,56 @@
+#include "c10/util/Exception.h"
+#include "MKLDNNTensorImpl.h"
+
+namespace c10 { namespace mkldnn {
+
+MKLDNNTensorImpl::MKLDNNTensorImpl(c10::TensorTypeId type_id, const caffe2::TypeMeta& data_type)
+  : c10::TensorImpl(type_id, data_type, nullptr, false) {}
+
+IntArrayRef MKLDNNTensorImpl::sizes() const {
+  return sizes_;
+}
+IntArrayRef MKLDNNTensorImpl::strides() const {
+  AT_ERROR("mkldnn tensors do not have strides");
+}
+bool MKLDNNTensorImpl::is_contiguous() const {
+  AT_ERROR("mkldnn tensors do not have is_contiguous");
+}
+int64_t MKLDNNTensorImpl::size(int64_t d) const {
+  return it_.get_dims()[d];
```

Comment:
why do you actually an MKLDNNTensorImpl?  Presumably we could support a number of "opaque" layouts and it's not obvious that they need separate TensorImpls.  They could support holding an opaque handle and we could support reading out the handle a number of different ways for actually implementing the kernels.

For example in this case, I don't see why we'd want to use the opaque handles size here instead of the size that is actually in TensorImpl.

Reply:
> why do you actually an MKLDNNTensorImpl? Presumably we could support a number of "opaque" layouts and it's not obvious that they need separate TensorImpls. They could support holding an opaque handle and we could support reading out the handle a number of different ways for actually implementing the kernels.

Are you suggesting to put an opaque handle directly into TensorImpl without creating new XYZTensorImpl classes? That should also work while we have to put some AT_ASSERT in methods of TensorImpl to avoid users calling into them accidentally on opaque layouts, similar to the situation if we want to make sparse tensor an opaque handle.

> For example in this case, I don't see why we'd want to use the opaque handles size here instead of the size that is actually in TensorImpl.

The problem is that the storage of opaque tensor is usually managed inside the opaque handle so is the dims/sizes info, e.g. MKL-DNN's memory descriptor. It is a headache to sync between TensorImpl::sizes_ and the one inside opaque handle and make sure TensorImpl::size is always valid. In that sense, I would rather do AT_ASSERT for the size() calls and allow MKL-DNN kernels to query opaque handle for size directly.

Reply:
> Are you suggesting to put an opaque handle directly into TensorImpl without creating new XYZTensorImpl classes?

Yes, this will then directly enable support of all other opaque layout tensor format (e.e.g opencl). For all these tensor layouts, what the pytorch core need to provide is the dispatching mechanism, and then specialized kernel will handle their own tensor layout. With different TensorTypeId should already be sufficient, no need for different tensorimpl.

> That should also work while we have to put some AT_ASSERT in methods of TensorImpl to avoid users calling into them accidentally on opaque layouts, 

Yes that's exactly what we want, basically add `AT_ASSERT(opaque_data_ == nullptr)` in all tensorimpl methods that assume dense plain format (e.g. slice)

> The problem is that the storage of opaque tensor is usually managed inside the opaque handle so is the dims/sizes info, e.g. MKL-DNN's memory descriptor. It is a headache to sync between TensorImpl::sizes_ and the one inside opaque handle and make sure TensorImpl::size is always valid.

I'm not sure whether it will be very hard, because for each mkldnn kernel, inputs tensors are supposed to be const, so the only thing you need to sync their sizes are the outputs tensor (with their underlying ideep::Tensor), which I imagine should be easy:
```
Tensor mkldnn_xxx(Tensor t1, Tensor t2) {
  ideep::Tensor i1 = static_cast<ideep::Tensor>t1.opaque_data();
  ideep::Tensor i2 = static_cast<ideep::Tensor>t2.opaque_data();
  // some mkldnn function calls
  ideep::Tensor o1 = ...
  return WrapIdeepTensor(o1);
}
```

>  I would rather do AT_ASSERT for the size() calls and allow MKL-DNN kernels to query opaque handle for size directly.

Since the plan is to expose mkldnn tensor to eager/python frontend, I believe allowing querying sizes (and some other meta info like dtype) would be very useful.

Reply:
@bddppq @gchanan Updated according to your comments. The PR description is also updated. Please review again. Thanks.